internal/backend/gs/gs.go

```diff
@@ -301,16 +317,45 @@ func (be *Backend) openReader(ctx context.Context, h restic.Handle, length int,
 
 	objName := be.Filename(h)
 
+	if be.encryptKey != "" {
+		encryptKeyPresent = true
+	}
+	for _, key := range be.decryptKeys {
+		if key != "" {
+			decryptKeysPresent = true
+			break
+		}
+	}
+
 	be.sem.GetToken()
 
-	r, err := be.bucket.Object(objName).NewRangeReader(ctx, offset, int64(length))
+	//https://cloud.google.com/storage/docs/encryption/using-customer-supplied-keys
+	err = errors.New(("Google Storage customer-supplied encryption keys failed."))
+	if encryptKeyPresent {
+		cryptoKey, _ = base64.StdEncoding.DecodeString(be.encryptKey)
+		rangeReader, err = be.bucket.Object(objName).Key(cryptoKey).NewRangeReader(ctx, offset, int64(length))
+	}
+	if decryptKeysPresent && err != nil {
+		for _, key := range be.decryptKeys {
+
+			cryptoKey, _ = base64.StdEncoding.DecodeString(key)
+			rangeReader, err = be.bucket.Object(objName).Key(cryptoKey).NewRangeReader(ctx, offset, int64(length))
```

Comment:
Doesn't this issue up to 100 API requests? The documentation states that

> When decrypting, gsutil calculates the SHA256 hash of the supplied encryption and decryption keys and selects the correct key to use for a particular object by matching the SHA256 hash in the object's metadata.

That sounds to me like it would be possible to determine which key will work.

Reply:
I hadn't considered that. Yeah, the metadata contains a sha of the key, `CustomerKeySHA256`, that can be compared locally. I'll add the sha calculations and base64 conversions to do the key selection locally. I'll make sure to avoid the metadata call if no keys are defined.

Reply:
Changed to calculate all the SHAs during the load of the config into a map. Then we simply use the string sha found on the object metadata to quickly dereference which loaded key to use to decrypt.